### ISSUE TITLE
Pass model/non-string POST parameters to controller?

### ISSUE THREAD

**Issue by AndrewDavid** [skipperbent/simple-php-router#438]
So, here's a super simple breakdown of what I'm trying to accomplish.

I have the following router:
`SimpleRouter::post('/user/login', 'User@login')->setName('userLogin');`

And the related method:
```php
public function login(User $user):User
{
   // Validation
   // Login using $user->username and $user->password
   // Etc.
   $user->id = $id;
   return $user;
}
```

And I'm trying to pass the POST data cast as the `User` model as a controller parameter.

I know I can fetch the POST data using `$request->getInputHandler()->all();` and then convert that to a `User` object, but that's not ideal for me. I also know I could use a Middleware of some sort that converts the POST data to a `User` object, store it in the request, then pass it to the controller in a callback like so:
```php
SimpleRouter::post('/user/login', function(){
   $userController = new User(); // or pass this via injection, whatever
   $user = $userController->login($request->user);
   return $user;
})->addMiddleware(UserConverter::class)->setName('userLogin');

// Where UserConverter basically populates $request->user
// with the POST data that was converted into a User object.
```

_(Code written roughly just to get the idea across)_

Is there a way to accomplish this? Am I just missing something? I've looked through the issues and the documentation but I just can't find anything for that. If this isn't currently possible with this library (which would be too bad, I really like this one so far), then are there any plans to implement that feature? Or is there another library you would recommend that works roughly the same but has what I need?

Thank you!

**Comment by AndrewDavid** [comment]
Ended up doing something like this:

```php
// Passing my php-di container ($container) using "use()"
SimpleRouter::post('/user/login', function() use ($container) {
  $userController = $container->get(\library\controllers\User::class);
  
  $postParameters = SimpleRouter::request()->getInputHandler()->all();

  // Using a static method called "cast" in the model to convert the POST array to an instance of the User model
  $user = \library\models\User::cast($postParameters);
  
  return $userController->login($user);
})->setName('userLogin');
```

This is going to become pretty repetitive if I have to write a function like this for each POST route however, so I'm open to suggestions/recommendations if you think there are better ways to handle this!

I know I could potentially write a `BaseController` that contains a routing method and have every other controller extend it. That way, all the POST routes can call that same method in every controller and the method itself can convert the POST parameters to the appropriate type based on the request. It would avoid the repetition, I suppose, but it still wouldn't be as ideal as I'd like it to be.

Let me know your thoughts!

Otherwise, feel free to close this issue. Thanks!

**Comment by AndrewDavid** [comment]
Quick update:

I figured out that I can use `setParameter` method to add a parameter to the route, like so:

```php
SimpleRouter::post('user/login', 'User@login')
  ->setParameters([
    'user' => \library\models\User::cast(SimpleRouter::request()->getInputHandler()->all())
  ])
  ->addMiddleware(UserConverter::class)
  ->setName('userLogin');
```

And thought I could get the Middleware to convert that parameter into GET parameters that the router will pass to the class, like so:

```php
public function handle(Request $request): void
{
  foreach($request->getLoadedRoute()->getParameters() as $index => $value)
    $request->getInputHandler()->addGet($index, $value);
}
```

But, as it turns out, the loadedRoute is only populated after the Middleware is processed, so I can't access the route parameters through there. On top of that, I discovered later on that the `addGet` method only accepts an object that implements `IInputItem` as its value (second argument). And the `IInputItem`'s value parameter can only be a string, not an object. So that basically threw me right back to square one.

Is there really absolutely no way to pass an object/model to the controller?

**Comment by skipperbent** [comment]
I know it's now exactly what you want, but you could populate the request object with your custom objects - the request object is available throughout the router.

```php
$request->key = $value;

// Or merge your array
$request += $values;
```

I'll make sure that the `IInputItem` value property will accept objects in the next release.

**Comment by skipperbent** [comment]
This issue has been resolved in version 4.3.0.0.
Feel free to open another issue if you are still experiencing issues.
